pinot-spi/src/main/java/org/apache/pinot/spi/executor/HardLimitExecutor.java

```diff
@@ -46,21 +46,28 @@ public HardLimitExecutor(int max, ExecutorService executorService) {
    */
   public static int getMultiStageExecutorHardLimit(PinotConfiguration config) {
     try {
-      int maxThreads = Integer.parseInt(config.getProperty(
+      int serverConfigLimit = config.getProperty(CommonConstants.Server.CONFIG_OF_MSE_MAX_EXECUTION_THREADS,
+              CommonConstants.Server.DEFAULT_MSE_MAX_EXECUTION_THREADS);
+      if (serverConfigLimit > 0) {
+        return serverConfigLimit;
+      }
+      int maxThreadsFromClusterConfig = Integer.parseInt(config.getProperty(
           CommonConstants.Helix.CONFIG_OF_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS,
           CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS
       ));
       int hardLimitFactor = Integer.parseInt(config.getProperty(
           CommonConstants.Helix.CONFIG_OF_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR,
           CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR
       ));
-      if (maxThreads <= 0 || hardLimitFactor <= 0) {
+      if (maxThreadsFromClusterConfig <= 0 || hardLimitFactor <= 0) {
         return 0;
       }
-      return maxThreads * hardLimitFactor;
+      return maxThreadsFromClusterConfig * hardLimitFactor;
     } catch (NumberFormatException e) {
-      return Integer.parseInt(CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS)
-          * Integer.parseInt(CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR);
+      int defaultLimitFromClusterConfig =
+              Integer.parseInt(CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS)
+              * Integer.parseInt(CommonConstants.Helix.DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR);
+      return Math.min(CommonConstants.Server.DEFAULT_MSE_MAX_EXECUTION_THREADS, defaultLimitFromClusterConfig);
```

Comment:
Not introduced in this PR, but it will always return negative value. Consider log a warning and return `-1` which disables the throttling